daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Ref.scala

```diff
@@ -27,64 +27,69 @@ object Ref {
     segments.result()
   }
 
-  case class DottedName private (segments: ImmArray[String]) {
-    override def toString: String = segments.toSeq.mkString(".")
-    def dottedName: String = toString
+  // We are very restrictive with regards to identifiers, taking inspiration
+  // from the lexical structure of Java:
+  // <https://docs.oracle.com/javase/specs/jls/se10/html/jls-3.html#jls-3.8>.
+  //
+  // In a language like C# you'll need to use some other unicode char for `$`.
+  val Name = MatchingStringModule("""[A-Za-z\$_][A-Za-z0-9\$_]*""".r)
+  type Name = Name.T
```

Comment:
```suggestion
  type Name = Name.T
  implicit def `Name equal instance`: Equal[Name] = Name.equalInstance
```

With `-Xsource:2.13` this will be discovered automatically; there is no need to import `equalInstance` at points of use.

Reply:
I am not sure what you mean.

Reply:
I add in `-XSource:2.13` in both 
`daml-lf/data/BUILD.bazel` and `daml-lf/scenario-interpreter/BUILD.bazel` and 
`daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/value/Value.scala` does not compile without the import of `Name equal instance`

Reply:
This _should_ work, but I might have an even better idea, and anyway you've addressed my main concern very nicely, so let's take it to a separate PR.